pkg/controller/autoscaling/elasticsearch/autoscaler/horizontal.go

```diff
@@ -49,20 +49,89 @@ func (ctx *Context) scaleHorizontally(
+// This function is only applicable to storage as it would otherwise prevent scale down of other resources.
+func (ctx *Context) getNodesToAddForStorage(
+	nodeCapacity resources.NodeResources,
+	requiredTotalStorageCapacity *client.AutoscalingCapacity,
+) int32 {
+	totalCurrentCapacity := ctx.AutoscalingPolicyResult.CurrentCapacity.Total // total capacity as observed by Elasticsearch
+	currentNodes := len(ctx.AutoscalingPolicyResult.CurrentNodes)
+	if totalCurrentCapacity.Storage.Value() > int64(currentNodes)*ctx.AutoscalingSpec.StorageRange.Max.Value() {
+		// The current storage capacity exceeds the maximum expected one. Since other other resources maybe scaled linearly
+		// according to the storage capacity it may lead to an ineffective scaling of other resources.
+		// See https://github.com/elastic/cloud-on-k8s/issues/4469
+		ctx.Log.Info(
+			"Current total storage capacity is greater than the one specified in the autoscaling specification.",
+			"policy", ctx.AutoscalingSpec.Name,
+			"scope", "tier",
+			"resource", "storage",
+			"current_total_storage_capacity", totalCurrentCapacity.Storage.Value(),
+			"max_storage_capacity_per_node", ctx.AutoscalingSpec.StorageRange.Max.Value(),
+			"current_node_count", currentNodes,
+		)
+
+		// Also surface this situation in the status.
+		ctx.StatusBuilder.
+			ForPolicy(ctx.AutoscalingSpec.Name).
+			RecordEvent(
+				status.UnexpectedStorageCapacity,
+				fmt.Sprintf(
+					"Current total storage capacity is %d, it is greater than the maximum expected one: %d (%d nodes * %d)",
+					totalCurrentCapacity.Storage.Value(),
+					int64(currentNodes)*ctx.AutoscalingSpec.StorageRange.Max.Value(),
+					currentNodes,
+					ctx.AutoscalingSpec.StorageRange.Max.Value(),
+				),
+			)
+	}
+
+	currentResources, hasStatus := ctx.CurrentAutoscalingStatus.CurrentResourcesForPolicy(ctx.AutoscalingSpec.Name)
+	if !hasStatus ||
+		requiredTotalStorageCapacity.Value() > totalCurrentCapacity.Storage.Value() {
+		// We are in one of the following situation:
+		// * The status is empty, this might happen if the autoscaling controller never ran on this cluster.
+		// * The total required capacity (at the policy level) is greater than the observed capacity, we should scale up.
+		nodeStorage := nodeCapacity.GetRequest(corev1.ResourceStorage)
+		return ctx.getNodesToAdd(
+			nodeStorage.Value(),
+			requiredTotalStorageCapacity.Value(),
+			string(corev1.ResourceStorage),
+		)
+	}
+
+	// reuse the actual number of nodes from the status
+	totalExpected := currentResources.NodeSetNodeCount.TotalNodeCount()
+	// we still want to downscale if user required it.
+	if totalExpected > ctx.AutoscalingSpec.NodeCountRange.Max {
+		totalExpected = ctx.AutoscalingSpec.NodeCountRange.Max
+	}
+	if totalExpected > ctx.AutoscalingSpec.NodeCountRange.Min {
+		return totalExpected - ctx.AutoscalingSpec.NodeCountRange.Min
+	}
```

Comment:
I am struggling to understand this logic here. I suppose the intend is to ensure that we downscale if the current capacity exceeds the specified maximum but I don't understand how that is achieved: 

Assuming min 1, max 3, total expected 2 would we not return 1 here which means adding another node, why?
Or min 1, max 3, total expected 4, return 2 i.e adding 2 nodes? I am almost certain I am missing something here but maybe we need a comment to clarify?

Reply:
You're right, it is not clean and misleading.
I'm working on a refactoring.